File: platform_cc/commands/all.py
"""
This file is part of Platform.CC.

Platform.CC is free software: you can redistribute it and/or modify
it under the terms of the GNU General Public License as published by
the Free Software Foundation, either version 3 of the License, or
(at your option) any later version.

Platform.CC is distributed in the hope that it will be useful,
but WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
GNU General Public License for more details.

You should have received a copy of the GNU General Public License
along with Platform.CC.  If not, see <https://www.gnu.org/licenses/>.
"""

from __future__ import absolute_import
import os
import time
import docker
from cleo import Command
from .common import outputJson, outputTable
from ..exception.state_error import StateError
from ..core.project import PlatformProject
from ..core.router import PlatformRouter
from ..application.base import BasePlatformApplication
from ..services.base import BasePlatformService

class AllStop(Command):
    """
    Stop all running projects.

    all:stop
    """

    def handle(self):
        projects = PlatformProject.getAllActiveProjects()
        for project in projects:
            projectContainers = project.dockerFetch(all=True)
            for container in projectContainers:
                container.stop()
            project.removeRouter()
        PlatformRouter().stop()

class AllPurge(Command):
    """
    Purge all Platform.CC related Docker contains, images, volumes, and networks.

    all:purge
        {--d|dry-run : List items to be purged.}
    """

    def handle(self):
        dryRun = bool(self.option("dry-run"))
        # inform user and wait 5 seconds
        if not dryRun:
            self.line(
                "<question>!!! Purge of ALL Platform.CC projects will commence in 10 seconds. Press CTRL+C to cancel. !!!</question>"
            )
            time.sleep(10)
        projects = PlatformProject.getAllActiveProjects()
        for project in projects:
            project.purge(dryRun)

class AllList(Command):
    """
    List details about all running projects.

    all:list
        {--j|json : If set output in JSON.}
    """

    def handle(self):
        projects = PlatformProject.getAllActiveProjects()
        outputData = []
        for project in projects:
            appList = []
            serviceList = []
            subnet = ""
            projectContainers = project.dockerFetch(all=True)
            for container in projectContainers:
                if isinstance(container, BasePlatformApplication):
                    appList.append({
                        "name" : container.getName(),
                        "type" : container.getType(),
                        "image" : container.getDockerImage()
                    })
                elif isinstance(container, BasePlatformService):
                    serviceList.append({
                        "name" : container.getName(),
                        "type" : container.getType(),
                        "image" : container.getDockerImage()
                    })
                if not subnet:
                    subnet = container.getNetwork().attrs.get("IPAM", {}).get("Config", [{}])[0].get("Subnet", "n/a")
            if not appList and not serviceList: continue
            outputData.append({
                "short_uid" : project.getShortUid(),
                "applications" : appList,
                "services" : serviceList,
                "subnet" : subnet
            })

        # json output
        if self.option("json"):
            outputJson(
                self,
                outputData
            )
            return
        
        # terminal tables output
        tableData = [
            ("Uid", "Applications", "Services", "Subnet"),
        ]
        for projectData in outputData:
            appNames = []
            serviceNames = []
            for app in projectData.get("applications", []):
                appNames.append("%s/%s" % (app["name"], app["type"]))
            for service in projectData.get("services", []):
                serviceNames.append("%s/%s" % (service["name"], service["type"]))
            tableData.append(
                (
                    projectData["short_uid"],
                    "\n".join(appNames).strip(),
                    "\n".join(serviceNames).strip(),
                    projectData["subnet"]
                )
            )
        outputTable(
            self,
            "Active Projects",
            tableData
        )

File: cmd/all.go
/*
This file is part of Platform.CC.

Platform.CC is free software: you can redistribute it and/or modify
it under the terms of the GNU General Public License as published by
the Free Software Foundation, either version 3 of the License, or
(at your option) any later version.

Platform.CC is distributed in the hope that it will be useful,
but WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
GNU General Public License for more details.

You should have received a copy of the GNU General Public License
along with Platform.CC.  If not, see <https://www.gnu.org/licenses/>.
*/

package cmd

import (
	"encoding/json"
	"fmt"
	"time"

	"github.com/spf13/cobra"
	"gitlab.com/contextualcode/platform_cc/api/container"
	"gitlab.com/contextualcode/platform_cc/api/router"
)

var allCmd = &cobra.Command{
	Use:   "all",
	Short: "Manage all projects.",
}

var allStopCmd = &cobra.Command{
	Use:   "stop",
	Short: "Stop all Platform.CC containers.",
	Run: func(cmd *cobra.Command, args []string) {
		containerHandler, err := getContainerHandler()
		handleError(err)
		handleError(containerHandler.AllStop())
	},
}

var allPurgeCmd = &cobra.Command{
	Use:   "purge",
	Short: "Purge all Platform.CC data.",
	Run: func(cmd *cobra.Command, args []string) {
		fmt.Println("!!! WARNING: PURGING ALL PLATFORM.CC DATA IN 5 SECONDS !!!")
		time.Sleep(5 * time.Second)
		containerHandler, err := getContainerHandler()
		handleError(err)
		handleError(containerHandler.AllPurge())
	},
}

var allStatusCmd = &cobra.Command{
	Use:     "status [--json]",
	Aliases: []string{"stat", "st"},
	Short:   "Show the status of all Platform.CC containers.",
	Run: func(cmd *cobra.Command, args []string) {
		containerHandler, err := getContainerHandler()
		handleError(err)
		// retrieve status
		stats, err := containerHandler.AllStatus()
		handleError(err)
		// json out
		jsonFlag := cmd.Flags().Lookup("json")
		if jsonFlag != nil && jsonFlag.Value.String() != "false" {
			out, err := json.Marshal(stats)
			handleError(err)
			fmt.Println(string(out))
			return
		}
		// table out
		data := make([][]string, 0)
		for _, s := range stats {
			ipAddrStr := "n/a"
			if s.IPAddress != "" {
				ipAddrStr = s.IPAddress
			}
			slot := "n/a"
			if s.Slot > 0 {
				slot = fmt.Sprintf("%d", s.Slot)
			}
			if s.ProjectID == "" && s.Name == "" && s.Image == router.GetContainerConfig().Image {
				s.ProjectID = "-"
				s.Name = "router"
				slot = "n/a"
				s.ObjectType = container.ObjectContainerRouter
			}
			serviceType := s.Type
			if s.Committed {
				serviceType = "[c] " + s.Type
			}
			data = append(data, []string{
				s.ProjectID,
				fmt.Sprintf("[%s] %s", string(s.ObjectType), s.Name),
				serviceType,
				slot,
				ipAddrStr,
			})
		}
		drawTable(
			[]string{"Project ID", "Name", "Type", "Slot", "IP Address"},
			data,
		)
		drawKeys()
		println("")
	},
}

func init() {
	allCmd.AddCommand(allStopCmd)
	allCmd.AddCommand(allPurgeCmd)
	allStatusCmd.Flags().Bool("json", false, "JSON output")
	allCmd.AddCommand(allStatusCmd)
	RootCmd.AddCommand(allCmd)
}
